### PR TITLE
Use decl_error in cennzx-spot

### PR DESCRIPTION
**Pull request by hoani**
- Change all Cennzx spot module errors into decl_error enum
- Decouple Cennzx Spot FeeRate Error type from Cennzx spot module

### CODE REVIEW

**Review by cowboy-bebug — APPROVED** [review]
LGTM nice work!

File: crml/cennzx-spot/src/lib.rs [hunk 1]
```diff
@@ -55,8 +56,83 @@ pub trait Trait: frame_system::Trait + pallet_generic_asset::Trait {
 	type UnsignedIntToBalance: From<LowPrecisionUnsigned> + Into<<Self as pallet_generic_asset::Trait>::Balance>;
 }
 
+decl_error! {
+	/// Error for the generic-asset module.
```
Comment (jordy25519) [inline crml/cennzx-spot/src/lib.rs]: Update module name

File: crml/cennzx-spot/src/types.rs [hunk 2]
```diff
@@ -53,18 +53,14 @@ impl Scaled for PerCent {
 
 // TODO: refactor below to use frame_support::decl_error!
```
Comment (jordy25519) [inline crml/cennzx-spot/src/types.rs]: Can remove the TODO

File: crml/cennzx-spot/src/lib.rs [hunk 3]
```diff
@@ -55,8 +56,83 @@ pub trait Trait: frame_system::Trait + pallet_generic_asset::Trait {
...
+	pub enum Error for Module<T: Trait> {
+		/// Exchange pool is empty.
+		EmptyExchangePool,
+		// Insufficient trade asset reserve in exchange
```
Comment (aliXsed) [inline crml/cennzx-spot/src/lib.rs:63]: `///` for all enum entries

File: crml/cennzx-spot/src/lib.rs [hunk 4]
```diff
@@ -55,8 +56,83 @@ pub trait Trait: frame_system::Trait + pallet_generic_asset::Trait {
...
+		// Minimum core asset is required
+		MinimumCoreAssetIsRequired,
+		// Assets withdrawn to be greater than zero
+		AssetToWithdrawNotAboveZero,
+		// Amount of exchange asset to burn should exist
+		LiquidityToWithdrawNotAboveZero,
+		// Liquidity should exist
+		NoLiquidityToRemove,
+		// trade asset amount must be greater than zero
+		TradeAssetToAddLiquidityNotAboveZero,
```
Comment (alexsednz) [inline crml/cennzx-spot/src/lib.rs]: Seems like we can merge a few of such errors in one enum entry as "AssetNotAboveZero". I know we want to generate a different error message, we should also take the fact that the type of the error is the same into account.

File: crml/cennzx-spot/src/lib.rs [hunk 5]
```diff
@@ -55,8 +56,83 @@ pub trait Trait: frame_system::Trait + pallet_generic_asset::Trait {
...
+		// core asset amount must be greater than zero
+		CoreAssetToAddLiquidityNotAboveZero,
+		// not enough core asset in balance
+		CoreAssetBalanceToAddLiquidityTooLow,
```
Comment (alexsednz) [inline crml/cennzx-spot/src/lib.rs]: The same here for "LiquidityTooLow"

**Review by jordy25519 — COMMENTED** [review]
Quick first pass review

File: crml/cennzx-spot/src/lib.rs [hunk 6]
```diff
@@ -55,8 +56,83 @@ pub trait Trait: frame_system::Trait + pallet_generic_asset::Trait {
...
+	/// Error for the generic-asset module.
+	pub enum Error for Module<T: Trait> {
+		/// Exchange pool is empty.
+		EmptyExchangePool,
+		// Insufficient trade asset reserve in exchange
+		InsufficientTradeAssetReserve,
+		// Insufficient core asset reserve in exchange
+		InsufficientCoreAssetReserve,
+		// Insufficient asset balance in buyer account
+		InsufficientBuyerTradeAssetBalance,
```
Comment (aliXsed) [inline crml/cennzx-spot/src/lib.rs:68]: I think we probably don't need that many error messages as the real type of errors is the same in many places. Similar to the Overflow error that we have used in many places regardless of where it has happened, we could have a few error types such as InsufficientAssetBalance, AssetNotAboveZero, AmountNotAboveZero,... 
However, as this is a refactor which is not pursuing any significant output change, I can see you may not want to go down that route.

Comment (hoani) [reply crml/cennzx-spot/src/lib.rs:68]: Yeah, there are as many error types as there were unique messages in the past. Totally agree that we should consider reducing the number of errors going forward

**Review by jordy25519 — APPROVED** [review]
LGTM!

**Pull request merged**